resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/StagedResourcesExpirationManager.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes
+
+import java.util.concurrent.{ScheduledExecutorService, TimeUnit}
+
+import scala.collection.JavaConverters._
+import scala.collection.concurrent.TrieMap
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.{Clock, Utils}
+
+private[spark] trait StagedResourcesExpirationManager {
+
+  def startMonitoringForExpiredResources(): Unit
+
+  def monitorResourceForExpiration(
+      resourceId: String,
+      podNamespace: String,
+      podLabels: Map[String, String],
+      kubernetesCredentials: PodMonitoringCredentials): Unit
+}
+
+private class StagedResourcesExpirationManagerImpl(
+      stagedResourcesStore: StagedResourcesStore,
+      kubernetesClientProvider: ResourceStagingServiceKubernetesClientProvider,
+      expirationExecutorService: ScheduledExecutorService,
+      clock: Clock,
+      expiredResourceTtlMs: Long,
+      resourceCleanupIntervalMs: Long)
+    extends StagedResourcesExpirationManager {
+
+  private val activeResources = TrieMap.empty[String, MonitoredResource]
+
+  override def startMonitoringForExpiredResources(): Unit = {
+    expirationExecutorService.scheduleAtFixedRate(
+        new CleanupRunnable(),
+        resourceCleanupIntervalMs,
+        resourceCleanupIntervalMs,
+        TimeUnit.MILLISECONDS)
+  }
+
+  override def monitorResourceForExpiration(
+      resourceId: String,
+      podNamespace: String,
+      podLabels: Map[String, String],
+      kubernetesCredentials: PodMonitoringCredentials): Unit = {
+    activeResources(resourceId) = MonitoredResource(
+        resourceId, podNamespace, podLabels, kubernetesCredentials)
+  }
+
+  private class CleanupRunnable extends Runnable with Logging {
+
+    private val expiringResources = mutable.Map.empty[String, ExpiringResource]
```

Comment:
why is this shared between runs?  to prevent double-deletes?  instead can we synchronize run, or otherwise if a run of the run method takes longer than the interval, skip the next run?